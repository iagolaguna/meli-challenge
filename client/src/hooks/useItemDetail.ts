import useSWR from 'swr'
import { fetcher } from 'utils/fetcher'

export const base = process.env.REACT_APP_API_BASE_PATH;
console.log(base)
export enum Condition {
    NEW = "new",
}

type useItemDetailProps = {
    id: string;
}

export type ItemDetail = {
    author: {
        name: string;
        lastname: string;
    },
    item: {
        id: string;
        title: string;
        price: {
            price: number;
            currency: string;
            amount: number;
        };
        picture: string;
        free_shipping: boolean;
        condition: Condition;
        description: string;
        sold_quantity: number;
    }
};

const useItemDetail = ({ id }: useItemDetailProps) => {
    const { data, error } = useSWR<ItemDetail>(`${base}/items/${id}`, fetcher)

    return {
        data,
        error,
        isLoading: !error && !data
    }
}

export default useItemDetail;