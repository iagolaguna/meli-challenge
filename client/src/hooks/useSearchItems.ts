import useSWR from 'swr'
import { fetcher } from 'utils/fetcher'

export const base = process.env.REACT_APP_API_BASE_PATH;
console.log(base)

type useSearchItemsProps = {
    query: string
}
export type SearchResult = {
    author: {
        name: string;
        lastname: string;
    },
    categories: string[];
    items: SearchItem[]
}
export type SearchItem = {
    id: string;
    title: string;
    price: {
        price: number;
        currency: string;
        amount: number;
    };
    picture: string;
    free_shipping: boolean;
    condition: string;
};

const useSearchItems = ({ query }: useSearchItemsProps) => {
    const { data, error } = useSWR<SearchResult>(query && `${base}/search?q=${query}`, fetcher)

    return {
        data,
        error,
        isLoading: !error && !data
    }
}

export default useSearchItems;