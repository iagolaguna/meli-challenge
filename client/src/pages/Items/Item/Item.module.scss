@import "styles/colors.scss";

.item {
  display: flex;
  flex-direction: column;
  padding: 16px;
  cursor: pointer;
}

.image {
  border-radius: 4px;
  width: 180px;
  height: 180px;
  object-fit: contain;
  margin-right: 16px;
}

.content {
  display: flex;
  flex-direction: column;
  margin: 16px 0px;
  h1 {
    color: $black;
    font-size: 18px;
    font-weight: normal;
  }
}

.price {
  display: flex;
  align-items: center;
  font-size: 24px;
  color: $black;
  span {
    margin-right: 8px;
  }
  margin-bottom: 32px;
}

.container {
  display: flex;
  flex-direction: row;
  margin-bottom: 8px;
}
