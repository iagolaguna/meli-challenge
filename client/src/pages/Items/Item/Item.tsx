import { Item as ItemType } from "hooks/useSearchItems";
import IcFreeShipping from "assets/ic_shipping.png";
import IcFreeShipping2x from "assets/ic_shipping@2x.png";
import style from "./Item.module.scss";
import Divider from "components/Divider/Divider";
import { formatAmount } from "utils/formatAmount";

type ItemProps = {
  hasDivider?: boolean;
  onClick: () => void;
} & ItemType;

export const Item = ({
  title,
  picture,
  price,
  free_shipping,
  hasDivider = true,
  onClick,
}: ItemProps) => (
  <div className={style.item} onClick={onClick} role="listitem">
    <div className={style.container}>
      <img className={style.image} src={picture} alt="Producto" />
      <div className={style.content}>
        <div className={style.price}>
          <span aria-label="Precio">$ {formatAmount(price.amount)}</span>
          {free_shipping && (
            <img
              src={IcFreeShipping}
              srcSet={`${IcFreeShipping} 1200w, ${IcFreeShipping2x} 1900w`}
              alt="EnvÃ­o gratis icon"
            />
          )}
        </div>
        <h1>{title}</h1>
      </div>
    </div>
    {hasDivider && <Divider />}
  </div>
);
