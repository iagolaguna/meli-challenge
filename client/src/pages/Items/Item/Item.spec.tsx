import { fireEvent, render } from "@testing-library/react";
import { Condition } from "hooks/useItemDetail";
import { Item as SearchItem } from "hooks/useSearchItems";
import { Item } from "pages/Items/Item/Item";

const itemMock: SearchItem = {
  id: "MLA901421848",
  title: "Apple iPhone 11 (64 Gb) - Negro",
  price: {
    currency: "ARS",
    amount: 154999,
    decimals: 0,
  },
  picture: "http://http2.mlstatic.com/D_656548-MLA46114829749_052021-I.jpg",
  free_shipping: true,
  condition: Condition.NEW,
};

describe("<Item/>", () => {
  it("should render free shipping icon if flag free_shipping be true", () => {
    const handleClick = jest.fn();
    const { queryByAltText } = render(
      <Item {...itemMock} onClick={handleClick} />
    );
    expect(queryByAltText(/Envío gratis icon/i)).toBeDefined();
  });

  it("should not render free shipping icon if flag free_shipping be false", () => {
    const item: SearchItem = {
      ...itemMock,
      free_shipping: false,
    };
    const handleClick = jest.fn();
    const { queryByAltText } = render(<Item {...item} onClick={handleClick} />);
    expect(queryByAltText(/Envío gratis icon/i)).toBeNull();
  });

  it("should render <Divider/> component if flag hasDivider be true", () => {
    const handleClick = jest.fn();
    const { container } = render(
      <Item {...itemMock} hasDivider onClick={handleClick} />
    );
    expect(container).toMatchSnapshot();
  });

  it("should not render <Divider/> component if flag hasDivider be false", () => {
    const handleClick = jest.fn();
    const { container } = render(<Item {...itemMock} onClick={handleClick} />);
    expect(container).toMatchSnapshot();
  });

  it("should call onClick when component be clicked", async () => {
    const handleClick = jest.fn();
    const { container } = render(<Item {...itemMock} onClick={handleClick} />);
    fireEvent.click(container.firstChild!);
    expect(handleClick).toHaveBeenCalled();
  });
});
