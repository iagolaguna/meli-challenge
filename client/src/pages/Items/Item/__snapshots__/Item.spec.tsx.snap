// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item/> should not render <Divider/> component if flag hasDivider be false 1`] = `
<div>
  <div
    class="item"
    role="listitem"
  >
    <div
      class="container"
    >
      <img
        alt="Producto"
        class="image"
        src="http://http2.mlstatic.com/D_656548-MLA46114829749_052021-I.jpg"
      />
      <div
        class="content"
      >
        <div
          class="price"
        >
          <span
            aria-label="Precio"
          >
            $ 
            154,999
          </span>
          <img
            alt="Envío gratis icon"
            src="ic_shipping.png"
            srcset="ic_shipping.png 1200w, ic_shipping@2x.png 1900w"
          />
        </div>
        <h1>
          Apple iPhone 11 (64 Gb) - Negro
        </h1>
      </div>
    </div>
    <div
      aria-hidden="false"
      class="divider"
    />
  </div>
</div>
`;

exports[`<Item/> should render <Divider/> component if flag hasDivider be true 1`] = `
<div>
  <div
    class="item"
    role="listitem"
  >
    <div
      class="container"
    >
      <img
        alt="Producto"
        class="image"
        src="http://http2.mlstatic.com/D_656548-MLA46114829749_052021-I.jpg"
      />
      <div
        class="content"
      >
        <div
          class="price"
        >
          <span
            aria-label="Precio"
          >
            $ 
            154,999
          </span>
          <img
            alt="Envío gratis icon"
            src="ic_shipping.png"
            srcset="ic_shipping.png 1200w, ic_shipping@2x.png 1900w"
          />
        </div>
        <h1>
          Apple iPhone 11 (64 Gb) - Negro
        </h1>
      </div>
    </div>
    <div
      aria-hidden="false"
      class="divider"
    />
  </div>
</div>
`;
